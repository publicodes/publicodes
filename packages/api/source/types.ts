import type Engine from 'publicodes'
import type { PublicodesExpression } from 'publicodes'

export type Expressions = PublicodesExpression | PublicodesExpression[]

export type Situation = Partial<Record<string, PublicodesExpression>>

export type NewEngine = (
	expressions: Expressions,
	situation?: Situation
) => Engine
