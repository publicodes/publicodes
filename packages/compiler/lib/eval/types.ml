open Core
open Utils
open Utils.Output
module Any = Utils.Uid.Make ()

type concrete_type = Number | String | Bool | Date
[@@deriving sexp, compare, show]

type naked_t = Concrete of concrete_type | Any of Any.t

and t = naked_t Pos.t UnionFind.elem

let any ~pos () : t = Any (Any.mk ()) |> Pos.mk ~pos |> UnionFind.make

let concrete ~pos typ : t = Concrete typ |> Pos.mk ~pos |> UnionFind.make

let concrete_to_str = function
  | Number ->
      "un nombre"
  | String ->
      "un texte"
  | Bool ->
      "un boolÃ©en (oui / non)"
  | Date ->
      "une date"

let unify (t1 : t) (t2 : t) =
  let typ1 = t1 |> UnionFind.get in
  let typ2 = t2 |> UnionFind.get in
  match (Pos.value typ1, Pos.value typ2) with
  | Any _, Any _ ->
      return (UnionFind.union t1 t2)
  | Any _, Concrete _ ->
      return (UnionFind.merge (fun _ b -> b) t1 t2)
  | Concrete _, Any _ ->
      return (UnionFind.merge (fun a _ -> a) t1 t2)
  | Concrete c1, Concrete c2 ->
      if [%compare.equal: concrete_type] c1 c2 |> not then
        (* Todo replace with a unique type_error, with the pos of the different arguments *)
        let code, message = Err.type_incoherence in
        fatal_error ~pos:(Pos.pos typ1) ~kind:`Type ~code
          ~labels:
            [ Pos.mk ~pos:(Pos.pos typ1)
                (Format.sprintf "est %s" (concrete_to_str c1))
            ; Pos.mk ~pos:(Pos.pos typ2)
                (Format.sprintf "est %s" (concrete_to_str c2)) ]
          message
      else return t1
