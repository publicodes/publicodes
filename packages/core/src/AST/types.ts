import { ArrondiNode } from '../mecanisms/arrondi'
import { BarèmeNode } from '../mecanisms/barème'
import { ConditionNode } from '../mecanisms/condition'
import { ContextNode } from '../mecanisms/contexte'
import { DuréeNode } from '../mecanisms/durée'
import { EstNonDéfiniNode } from '../mecanisms/est'
import { EstNonApplicableNode } from '../mecanisms/est-non-applicable'
import { GrilleNode } from '../mecanisms/grille'
import { InversionNode } from '../mecanisms/inversion'
import { PossibilityNode } from '../mecanisms/one-possibility'
import { OperationNode } from '../mecanisms/operation'
import { RésoudreRéférenceCirculaireNode } from '../mecanisms/résoudre-référence-circulaire'
import { SimplifierUnitéNode } from '../mecanisms/simplifier-unité'
import { TauxProgressifNode } from '../mecanisms/tauxProgressif'
import { TexteNode } from '../mecanisms/texte'
import { UnitéNode } from '../mecanisms/unité'
import { VariableManquanteNode } from '../mecanisms/variablesManquantes'
import { VariationNode } from '../mecanisms/variations'
import { ReferenceNode } from '../reference'
import { ReplacementRule } from '../replacement'
import { RuleNode } from '../rule'

export type ConstantNode = {
	type: 'boolean' | 'number' | 'string' | undefined
	nodeValue: Evaluation
	nodeKind: 'constant'
	isNullable?: boolean
	isDefault?: boolean
}
type PossibleNodes =
	| RuleNode
	| ReferenceNode
	| ArrondiNode
	| BarèmeNode
	| DuréeNode
	| GrilleNode
	| EstNonApplicableNode
	| EstNonDéfiniNode
	| InversionNode
	| OperationNode
	| PossibilityNode
	| ContextNode
	| SimplifierUnitéNode
	| RésoudreRéférenceCirculaireNode
	| TauxProgressifNode
	| UnitéNode
	| VariationNode
	| ConditionNode
	| ConstantNode
	| ReplacementRule
	| VariableManquanteNode
	| TexteNode

export type NodeKind = PossibleNodes['nodeKind']
export type ASTNode<N extends NodeKind = NodeKind> = PossibleNodes & {
	nodeKind: N
	isDefault?: boolean
	sourceMap?: {
		mecanismName: string
		args: Record<string, ASTNode | Array<ASTNode>>
	}
	rawNode?: string | Record<string, unknown>
} & (
		| EvaluationDecoration<Types>
		// We remove the ESLINT warning as it does not concern intersection type and is actually useful here
		// https://github.com/typescript-eslint/typescript-eslint/issues/2063#issuecomment-675156492
		// eslint-disable-next-line @typescript-eslint/ban-types
		| {}
	)

// TODO : separate type for evaluated AST Tree

export type MecanismNode = ASTNode<
	Exclude<NodeKind, 'rule' | 'reference' | 'constant'>
>

export type ASTTransformer = (n: ASTNode) => ASTNode
export type ASTVisitor = (n: ASTNode) => void

export type TraverseFunction<K extends NodeKind> = (
	fn: ASTTransformer,
	node: ASTNode<K>,
) => ASTNode<K>

export type BaseUnit = string

// TODO: I believe it would be more effecient (for unit conversion and for
// inference), and more general to represent units using a map of base unit to
// their power number :
//
// type Unit = Map<BaseUnit, number>
// N.m²/kg² <-> {N: 1, m: 2, kg: -2} (gravity constant)
export type Unit = {
	numerators: Array<BaseUnit>
	denominators: Array<BaseUnit>
}

// Idée : une évaluation est un n-uple : (value, unit, missingVariables, isApplicable)
type EvaluationDecoration<T extends Types> = {
	nodeValue: Evaluation<T>
	unit?: Unit
	traversedVariables?: Array<string>
	missingVariables: MissingVariables
}
export type Types = number | boolean | string | Record<string, unknown>

export type Evaluation<T extends Types = Types> =
	| T
	| null // Non applicable
	| undefined // Non défini

export type EvaluatedNode<
	K extends NodeKind = NodeKind,
	T extends Types = Types,
> = EvaluationDecoration<T> & ASTNode<K>

export type MissingVariables = Record<string, number>
