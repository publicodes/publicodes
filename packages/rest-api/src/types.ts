import type PublicodesEngine from 'publicodes'
import type { PublicodesExpression } from 'publicodes'

export type Expressions = PublicodesExpression | PublicodesExpression[]

export type Situation = Partial<Record<string, PublicodesExpression>>

export type Engine = PublicodesEngine
