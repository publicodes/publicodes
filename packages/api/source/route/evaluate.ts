import { Expressions, NewEngine, Situation } from '../types.js'
import { catchError, PickInObject } from '../utils.js'

export interface EvaluateBody {
	expressions: Expressions
	situation?: Situation
}

export function evaluate(
	newEngine: NewEngine,
	{ expressions, situation }: EvaluateBody
) {
	const engine = newEngine(expressions, situation)
	const [error] = catchError(() => engine.setSituation(situation))

	const keysKept = [
		'nodeValue' as const,
		'unit' as const,
		'traversedVariables' as const,
		'missingVariables' as const,
	]

	const evaluateResult = (
		Array.isArray(expressions) ? expressions : [expressions]
	).map((expression) => {
		const [error, result] = catchError(() =>
			PickInObject(engine.evaluate(expression), keysKept)
		)

		return !error ? result : { error: { message: error.message } }
	})

	return {
		evaluate: evaluateResult,
		situationError: error && { message: error.message },
	}
}
