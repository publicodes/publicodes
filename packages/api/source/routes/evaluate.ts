import { Situation, NewEngine, Expressions } from '@/types'
import { catchError, PickInObject } from '@/utils'

export interface APIParameters {
	situation?: Situation
}

export function evaluate(
	newEngine: NewEngine,
	expressions: Expressions,
	{ situation }: APIParameters = {}
) {
	const engine = newEngine(expressions, situation)
	const situationResult = catchError(() => engine.setSituation(situation))

	const keysKept = [
		'nodeValue' as const,
		'unit' as const,
		'traversedVariables' as const,
		'missingVariables' as const,
	]

	const evaluateResult = (
		Array.isArray(expressions) ? expressions : [expressions]
	).map((expression) =>
		catchError(() => PickInObject(engine.evaluate(expression), keysKept))
	)

	return {
		situation: situationResult,
		evaluate: evaluateResult,
	}
}
