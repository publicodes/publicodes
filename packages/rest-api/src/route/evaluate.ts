import { Engine, Expressions, Situation } from '../types.js'
import { catchError, PickInObject } from '../utils.js'

export interface EvaluateBody {
	expressions: Expressions
	situation?: Situation
}

export function evaluate(
	originalEngine: Engine,
	{ expressions, situation }: EvaluateBody,
) {
	const engine = originalEngine.shallowCopy()
	originalEngine.subEngines = [] // This line avoid memory leak cause by multiple call to shallowCopy(), issue https://github.com/betagouv/publicodes/issues/239

	const warnings: Array<{ message: string }> = []
	if (engine.baseContext) {
		engine.baseContext.logger = {
			error: console.error,
			log: console.log,
			warn(warning: string) {
				warnings.push({ message: warning })
			},
		}
	}

	const [situationError] = catchError(() => engine.setSituation(situation))

	if (situationError) {
		return { situationError: { message: situationError.message } }
	}

	const keysKept = [
		'nodeValue' as const,
		'unit' as const,
		'traversedVariables' as const,
		'missingVariables' as const,
	]

	const evaluateResult = (
		Array.isArray(expressions) ? expressions : [expressions]).map(
		(expression) => {
			const [error, result] = catchError(() =>
				PickInObject(engine.evaluate(expression), keysKept),
			)

			return !error ? result : { error: { message: error.message }, warnings }
		},
	)

	return { evaluate: evaluateResult, warnings }
}
