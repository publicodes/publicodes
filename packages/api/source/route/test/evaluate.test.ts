import { Expressions, NewEngine, Situation } from '../../types'
import Engine from 'publicodes'
import { describe, expect, it, vi } from 'vitest'
import { evaluate } from '../evaluate'

const obj = {
	setSituation: (_situation?: Situation) => 42,
	evaluate: (_expressions: Expressions) => ({
		nodeValue: 42,
		unit: 42,
		traversedVariables: 42,
		missingVariables: 42,
	}),
}

const mockedNewEngine = ((_expressions: Expressions, _situation?: Situation) =>
	obj) as unknown as NewEngine

const rules = `
prix:
prix . carottes: 2€/kg
prix . champignons: 5€/kg
prix . avocat: 2€/avocat

dépenses primeur:
  formule:
    somme:
      - prix . carottes * 1.5 kg
      - prix . champignons * 500g
      - prix . avocat * 3 avocat
`

const newEngine = () => new Engine(rules)

describe('evaluate', () => {
	it('Test input/output', () => {
		const spySitu = vi.spyOn(obj, 'setSituation')
		const spyEval = vi.spyOn(obj, 'evaluate')

		expect(
			evaluate(mockedNewEngine, {
				expressions: 'expressions',
				situation: { test: 'test' },
			})
		).toMatchInlineSnapshot(`
			{
			  "evaluate": [
			    {
			      "missingVariables": 42,
			      "nodeValue": 42,
			      "traversedVariables": 42,
			      "unit": 42,
			    },
			  ],
			  "situationError": null,
			}
		`)

		expect(spySitu).toHaveBeenCalledWith({ test: 'test' })
		expect(spyEval).toHaveBeenCalledWith('expressions')
	})

	it('One expression in array should return same result as a single same expression', () => {
		expect(
			evaluate(mockedNewEngine, { expressions: ['coucou'] })
		).toMatchObject(evaluate(mockedNewEngine, { expressions: 'coucou' }))
	})

	it('Simple test with real Engine', () => {
		expect(evaluate(newEngine, { expressions: ['21 + 21', '6 * 7'] }))
			.toMatchInlineSnapshot(`
				{
				  "evaluate": [
				    {
				      "missingVariables": {},
				      "nodeValue": 42,
				      "traversedVariables": [],
				      "unit": undefined,
				    },
				    {
				      "missingVariables": {},
				      "nodeValue": 42,
				      "traversedVariables": [],
				      "unit": undefined,
				    },
				  ],
				  "situationError": null,
				}
			`)
	})

	it('Test with real Engine', () => {
		expect(
			evaluate(newEngine, {
				expressions: ['dépenses primeur', 'prix', 'prix . avocat * 21'],
			})
		).toMatchInlineSnapshot(`
			{
			  "evaluate": [
			    {
			      "missingVariables": {},
			      "nodeValue": 11.5,
			      "traversedVariables": [
			        "dépenses primeur",
			        "prix . carottes",
			        "prix . champignons",
			        "prix . avocat",
			      ],
			      "unit": {
			        "denominators": [],
			        "numerators": [
			          "€",
			        ],
			      },
			    },
			    {
			      "missingVariables": {
			        "prix": 1,
			      },
			      "nodeValue": undefined,
			      "traversedVariables": [
			        "prix",
			      ],
			      "unit": undefined,
			    },
			    {
			      "missingVariables": {},
			      "nodeValue": 42,
			      "traversedVariables": [
			        "prix . avocat",
			      ],
			      "unit": {
			        "denominators": [
			          "avocat",
			        ],
			        "numerators": [
			          "€",
			        ],
			      },
			    },
			  ],
			  "situationError": null,
			}
		`)
	})

	it('Test invalid syntax in situation', () => {
		expect(
			evaluate(newEngine, {
				expressions: ['1 + 1'],
				situation: { test: '"42' },
			})
		).toMatchSnapshot()
	})

	it('Test syntax error in expression', () => {
		expect(evaluate(newEngine, { expressions: '1+1' })).toMatchSnapshot()
	})

	it('Test two syntax error in expression', () => {
		expect(
			evaluate(newEngine, {
				expressions: ['1+1', '"42', '42'],
			})
		).toMatchSnapshot()
	})

	it('Test error in expression and situation at same time', () => {
		expect(
			evaluate(newEngine, { expressions: '1+1', situation: { test: '"42' } })
		).toMatchSnapshot()
	})

	it('Test no expressions', () => {
		expect(evaluate(newEngine, { expressions: [] })).toMatchInlineSnapshot(`
			{
			  "evaluate": [],
			  "situationError": null,
			}
		`)
	})
})
