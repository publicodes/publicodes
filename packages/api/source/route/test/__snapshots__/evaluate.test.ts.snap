// Vitest Snapshot v1

exports[`evaluate > Test error in expression and situation at same time 1`] = `
{
  "evaluate": [
    {
      "error": {
        "message": "
[ Erreur syntaxique ]
➡️  Dans la règle \\"evaluation\\"
✖️  \`1+1\` n'est pas une expression valide
    Syntax error at line 1 col 2:

  1+
   ^
Unexpected additionSubstraction token: \\"+\\". Instead, I was expecting to see one of the following:

A space token based on:
    number$ebnf$1$subexpression$1 →  ● %space
    number$ebnf$1 →  ● number$ebnf$1$subexpression$1
    number → %number ● number$ebnf$1 Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    MultiplicationDivision → MultiplicationDivision ● %space %multiplicationDivision %space Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A words token based on:
    UnitNumerator →  ● %words UnitNumerator$ebnf$1
    Unit$ebnf$1 →  ● UnitNumerator
    Unit →  ● Unit$ebnf$1 Unit$ebnf$2
    number → %number number$ebnf$1 ● Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    AdditionSubstraction → AdditionSubstraction ● %space %additionSubstraction %space MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    Comparison → Comparable ● %space %comparison %space Comparable
    main →  ● Comparison
A space token based on:
    UnitDenominator$ebnf$1$subexpression$1 →  ● %space
    UnitDenominator$ebnf$1 →  ● UnitDenominator$ebnf$1$subexpression$1
    UnitDenominator →  ● UnitDenominator$ebnf$1 \\"/\\" %words
    Unit$ebnf$2 → Unit$ebnf$2 ● UnitDenominator
    Unit → Unit$ebnf$1 ● Unit$ebnf$2
    number → %number number$ebnf$1 ● Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A \\"/\\" based on:
    UnitDenominator → UnitDenominator$ebnf$1 ● \\"/\\" %words
    Unit$ebnf$2 → Unit$ebnf$2 ● UnitDenominator
    Unit → Unit$ebnf$1 ● Unit$ebnf$2
    number → %number number$ebnf$1 ● Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    MultiplicationDivision → MultiplicationDivision ● %space %multiplicationDivision %space Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    AdditionSubstraction → AdditionSubstraction ● %space %additionSubstraction %space MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    Comparison → Comparable ● %space %comparison %space Comparable
    main →  ● Comparison

",
      },
    },
  ],
  "situationError": {
    "message": "
[ Erreur syntaxique ]
➡️  Dans la règle \\"situation [test]\\"
✖️  \`\\"42\` n'est pas une expression valide
    invalid syntax at line 1 col 1:

  \\"42
  ^
Unexpected input (lexer error). Instead, I was expecting to see one of the following:

A date token based on:
    Date →  ● %date
    main →  ● Date
A JSONObject token based on:
    JSONObject →  ● %JSONObject
    main →  ● JSONObject
A \\"-\\" based on:
    Negation →  ● \\"-\\" %space Parentheses
    NumericValue →  ● Negation
    main →  ● NumericValue
A words token based on:
    Variable →  ● %words Variable$ebnf$1
    Date →  ● Variable
    main →  ● Date
A \\"oui\\" based on:
    boolean →  ● \\"oui\\"
    NonNumericTerminal →  ● boolean
    main →  ● NonNumericTerminal
A \\"non\\" based on:
    boolean →  ● \\"non\\"
    NonNumericTerminal →  ● boolean
    main →  ● NonNumericTerminal
A string token based on:
    string →  ● %string
    NonNumericTerminal →  ● string
    main →  ● NonNumericTerminal
A \\"(\\" based on:
    Parentheses →  ● \\"(\\" %space NumericValue %space \\")\\"
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A \\"(\\" based on:
    Parentheses →  ● \\"(\\" NumericValue \\")\\"
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A number token based on:
    number →  ● %number
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A infinity token based on:
    number →  ● %infinity
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A number token based on:
    number →  ● %number number$ebnf$1 Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue

",
  },
}
`;

exports[`evaluate > Test invalid syntax in situation 1`] = `
{
  "evaluate": [
    {
      "missingVariables": [],
      "nodeValue": 2,
      "traversedVariables": [],
      "unit": undefined,
    },
  ],
  "situationError": {
    "message": "
[ Erreur syntaxique ]
➡️  Dans la règle \\"situation [test]\\"
✖️  \`\\"42\` n'est pas une expression valide
    invalid syntax at line 1 col 1:

  \\"42
  ^
Unexpected input (lexer error). Instead, I was expecting to see one of the following:

A date token based on:
    Date →  ● %date
    main →  ● Date
A JSONObject token based on:
    JSONObject →  ● %JSONObject
    main →  ● JSONObject
A \\"-\\" based on:
    Negation →  ● \\"-\\" %space Parentheses
    NumericValue →  ● Negation
    main →  ● NumericValue
A words token based on:
    Variable →  ● %words Variable$ebnf$1
    Date →  ● Variable
    main →  ● Date
A \\"oui\\" based on:
    boolean →  ● \\"oui\\"
    NonNumericTerminal →  ● boolean
    main →  ● NonNumericTerminal
A \\"non\\" based on:
    boolean →  ● \\"non\\"
    NonNumericTerminal →  ● boolean
    main →  ● NonNumericTerminal
A string token based on:
    string →  ● %string
    NonNumericTerminal →  ● string
    main →  ● NonNumericTerminal
A \\"(\\" based on:
    Parentheses →  ● \\"(\\" %space NumericValue %space \\")\\"
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A \\"(\\" based on:
    Parentheses →  ● \\"(\\" NumericValue \\")\\"
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A number token based on:
    number →  ● %number
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A infinity token based on:
    number →  ● %infinity
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A number token based on:
    number →  ● %number number$ebnf$1 Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue

",
  },
}
`;

exports[`evaluate > Test syntax error in expression 1`] = `
{
  "evaluate": [
    {
      "error": {
        "message": "
[ Erreur syntaxique ]
➡️  Dans la règle \\"evaluation\\"
✖️  \`1+1\` n'est pas une expression valide
    Syntax error at line 1 col 2:

  1+
   ^
Unexpected additionSubstraction token: \\"+\\". Instead, I was expecting to see one of the following:

A space token based on:
    number$ebnf$1$subexpression$1 →  ● %space
    number$ebnf$1 →  ● number$ebnf$1$subexpression$1
    number → %number ● number$ebnf$1 Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    MultiplicationDivision → MultiplicationDivision ● %space %multiplicationDivision %space Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A words token based on:
    UnitNumerator →  ● %words UnitNumerator$ebnf$1
    Unit$ebnf$1 →  ● UnitNumerator
    Unit →  ● Unit$ebnf$1 Unit$ebnf$2
    number → %number number$ebnf$1 ● Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    AdditionSubstraction → AdditionSubstraction ● %space %additionSubstraction %space MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    Comparison → Comparable ● %space %comparison %space Comparable
    main →  ● Comparison
A space token based on:
    UnitDenominator$ebnf$1$subexpression$1 →  ● %space
    UnitDenominator$ebnf$1 →  ● UnitDenominator$ebnf$1$subexpression$1
    UnitDenominator →  ● UnitDenominator$ebnf$1 \\"/\\" %words
    Unit$ebnf$2 → Unit$ebnf$2 ● UnitDenominator
    Unit → Unit$ebnf$1 ● Unit$ebnf$2
    number → %number number$ebnf$1 ● Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A \\"/\\" based on:
    UnitDenominator → UnitDenominator$ebnf$1 ● \\"/\\" %words
    Unit$ebnf$2 → Unit$ebnf$2 ● UnitDenominator
    Unit → Unit$ebnf$1 ● Unit$ebnf$2
    number → %number number$ebnf$1 ● Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    MultiplicationDivision → MultiplicationDivision ● %space %multiplicationDivision %space Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    AdditionSubstraction → AdditionSubstraction ● %space %additionSubstraction %space MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    Comparison → Comparable ● %space %comparison %space Comparable
    main →  ● Comparison

",
      },
    },
  ],
  "situationError": null,
}
`;

exports[`evaluate > Test two syntax error in expression 1`] = `
{
  "evaluate": [
    {
      "error": {
        "message": "
[ Erreur syntaxique ]
➡️  Dans la règle \\"evaluation\\"
✖️  \`1+1\` n'est pas une expression valide
    Syntax error at line 1 col 2:

  1+
   ^
Unexpected additionSubstraction token: \\"+\\". Instead, I was expecting to see one of the following:

A space token based on:
    number$ebnf$1$subexpression$1 →  ● %space
    number$ebnf$1 →  ● number$ebnf$1$subexpression$1
    number → %number ● number$ebnf$1 Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    MultiplicationDivision → MultiplicationDivision ● %space %multiplicationDivision %space Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A words token based on:
    UnitNumerator →  ● %words UnitNumerator$ebnf$1
    Unit$ebnf$1 →  ● UnitNumerator
    Unit →  ● Unit$ebnf$1 Unit$ebnf$2
    number → %number number$ebnf$1 ● Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    AdditionSubstraction → AdditionSubstraction ● %space %additionSubstraction %space MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    Comparison → Comparable ● %space %comparison %space Comparable
    main →  ● Comparison
A space token based on:
    UnitDenominator$ebnf$1$subexpression$1 →  ● %space
    UnitDenominator$ebnf$1 →  ● UnitDenominator$ebnf$1$subexpression$1
    UnitDenominator →  ● UnitDenominator$ebnf$1 \\"/\\" %words
    Unit$ebnf$2 → Unit$ebnf$2 ● UnitDenominator
    Unit → Unit$ebnf$1 ● Unit$ebnf$2
    number → %number number$ebnf$1 ● Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A \\"/\\" based on:
    UnitDenominator → UnitDenominator$ebnf$1 ● \\"/\\" %words
    Unit$ebnf$2 → Unit$ebnf$2 ● UnitDenominator
    Unit → Unit$ebnf$1 ● Unit$ebnf$2
    number → %number number$ebnf$1 ● Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    MultiplicationDivision → MultiplicationDivision ● %space %multiplicationDivision %space Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    AdditionSubstraction → AdditionSubstraction ● %space %additionSubstraction %space MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    Comparison → Comparable ● %space %comparison %space Comparable
    main →  ● Comparison

",
      },
    },
    {
      "error": {
        "message": "
[ Erreur syntaxique ]
➡️  Dans la règle \\"evaluation\\"
✖️  \`\\"42\` n'est pas une expression valide
    invalid syntax at line 1 col 1:

  \\"42
  ^
Unexpected input (lexer error). Instead, I was expecting to see one of the following:

A date token based on:
    Date →  ● %date
    main →  ● Date
A JSONObject token based on:
    JSONObject →  ● %JSONObject
    main →  ● JSONObject
A \\"-\\" based on:
    Negation →  ● \\"-\\" %space Parentheses
    NumericValue →  ● Negation
    main →  ● NumericValue
A words token based on:
    Variable →  ● %words Variable$ebnf$1
    Date →  ● Variable
    main →  ● Date
A \\"oui\\" based on:
    boolean →  ● \\"oui\\"
    NonNumericTerminal →  ● boolean
    main →  ● NonNumericTerminal
A \\"non\\" based on:
    boolean →  ● \\"non\\"
    NonNumericTerminal →  ● boolean
    main →  ● NonNumericTerminal
A string token based on:
    string →  ● %string
    NonNumericTerminal →  ● string
    main →  ● NonNumericTerminal
A \\"(\\" based on:
    Parentheses →  ● \\"(\\" %space NumericValue %space \\")\\"
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A \\"(\\" based on:
    Parentheses →  ● \\"(\\" NumericValue \\")\\"
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A number token based on:
    number →  ● %number
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A infinity token based on:
    number →  ● %infinity
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A number token based on:
    number →  ● %number number$ebnf$1 Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue

",
      },
    },
    {
      "missingVariables": [],
      "nodeValue": 42,
      "traversedVariables": [],
      "unit": undefined,
    },
  ],
  "situationError": null,
}
`;
